source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,413 @@
+
+    /// As defined in the Whitepaper this is the total rewards to be sent to
+    /// Commons Budget over about 6.7 years
+    private const Amount totalCommonsBudgetRewards = 1_800_000_000.coins;
+
+    /// As defined in the Whitepaper this is the total rewards allocated to be
+    /// distributed to all the Validators over about 128 years
+    private const Amount totalValRewards = 2_700_000_000.coins;
+
+    /// Every 5 secs 50 coins are rewarded equals 10 per second (Block interval of 600 secs)
+    private const Amount commonsCoinsPerBlock = 6000.coins;
+
+    /// This is the Amount to pay the Commons Budget in a payout period before the allocation runs out
+    private Amount commonsPerPayout;
+
+    /// Number of blocks that rewards are paid for each payout (also the interval for payments)
+    private uint payout_period;
+
+    /// Duration of a block in seconds
+    private ulong block_interval_sec;
+
+    /// Number of Coinbase transactions per year
+    private ulong payouts_per_year;
+
+    /***************************************************************************
+
+        Construct `Reward` object which encapsulates block reward calculations.
+
+        Params:
+            payout_period = how many blocks are included for a reward payout
+            block_interval_sec = time span for each block
+
+    ***************************************************************************/
+
+    public this (in uint payout_period, in Duration block_interval)
+    {
+        this.log = Logger(__MODULE__);
+
+        this.payout_period = payout_period;
+        this.block_interval_sec = block_interval.total!"seconds";
+
+        // Calculate payouts per year
+        this.payouts_per_year = yearOfSecs / (this.payout_period * this.block_interval_sec);
+
+        // Calculate amount that would be paid to Commons Budget for full payout period
+        // if the allocation is still sufficient
+        this.commonsPerPayout = commonsCoinsPerBlock; commonsPerPayout.mul(this.payout_period);
+    }
+
+    /***************************************************************************
+
+        Returns the Block Rewards for the given height.
+
+        Params:
+            height = last block of the current payout period
+                (the CoinBase transaction will be in block at this height plus the payout period)
+            percent_signed = (#actual signatures / #expected signatures) * 100
+                on all the blocks for payout period
+
+        Returns:
+            `BlockRewards` which indicates the `Amount` to pay the `Validators`
+            and the `Commons Budget`
+
+    ***************************************************************************/
+
+    public BlockRewards calculateBlockRewards (in Height height, in ubyte percent_signed) nothrow @safe
+    {
+        assert(height % this.payout_period == 0, "This is not a payout block height");
+        assert (height > 0, "We can not payout in Genesis block");
```

Comment:
Spurious space

Reply:
BTW another way to approach this is to define that there is no reward given to commons / validators at non-payout height. That avoids the assertion and it's still correct from the POV of this component.

Reply:
Updated